crates/rethnet_eth/src/remote/client.rs

```diff
@@ -226,67 +300,148 @@ impl RpcClient {
-    where
-        T: for<'a> serde::Deserialize<'a>,
-    {
-        let request = self.serialize_request(input);
+    async fn call_without_cache<T: DeserializeOwned>(
+        &self,
+        method_invocation: MethodInvocation,
+    ) -> Result<T, RpcClientError> {
+        let request = self.serialize_request(method_invocation)?;
 
-        self.send_request_body(&request.request)
+        self.send_request_body(&request)
             .await
-            .and_then(|response| Self::extract_response(request, response))
+            .and_then(|response| Self::extract_result(request, response))
     }
 
+    /// Returns the results of the given method invocations.
     async fn batch_call(
         &self,
-        inputs: &[MethodInvocation],
-    ) -> Result<BatchResponse, RpcClientError> {
-        let cache_key = CacheableMethodInvocations::try_from(inputs)
-            .ok()
-            .and_then(|v| v.cache_key());
+        method_invocations: &[MethodInvocation],
+    ) -> Result<VecDeque<serde_json::Value>, RpcClientError> {
+        let ids = self.get_ids(method_invocations.len() as u64);
 
-        let request_strings: Vec<String> = inputs
+        let cache_keys = method_invocations
             .iter()
-            .map(|i| self.serialize_request(i).request)
-            .collect();
+            .map(|m| {
+                CacheableMethodInvocation::try_from(m)
+                    .ok()
+                    .and_then(|v| v.cache_key())
+            })
+            .collect::<Vec<_>>();
 
-        let request_body = format!("[{}]", request_strings.join(","));
+        let mut results: Vec<Option<serde_json::Value>> =
+            Vec::with_capacity(method_invocations.len());
+        for cache_key in cache_keys.iter() {
+            let result = self.try_from_cache(cache_key.as_ref()).await?;
+            results.push(result)
+        }
 
-        self.send_request_body_with_cache(&request_body, cache_key)
-            .await
-            .map(|response| BatchResponse {
-                text: response,
-                request_strings,
+        let mut requests: Vec<SerializedRequest> = Vec::new();
+        for (id, method_invocation, cache_response) in izip!(&ids, method_invocations, &results) {
+            if cache_response.is_none() {
+                let request =
+                    Self::serialize_request_with_id(method_invocation.clone(), id.clone())?;
+                requests.push(request);
+            }
+        }
+
+        let request_body = SerializedRequest(
+            serde_json::to_value(&requests).map_err(RpcClientError::InvalidJsonRequest)?,
+        );
+        let remote_response = self.send_request_body(&request_body).await?;
+        let remote_responses: Vec<jsonrpc::Response<serde_json::Value>> =
+            Self::parse_response_str(&remote_response)?;
+
+        for response in remote_responses {
+            let index = ids
+                .iter()
+                .position(|id| id == &response.id)
```

Comment:
Currently we have `n = 3`, so the quadratic complexity here is fine, but we might wanna fetch a lot of blocks at once in batch requests later on, so I'll rewrite this tomorrow to use a better data structure.